python/tests/test_mixed_log_reader.py

```diff
@@ -0,0 +1,405 @@
+import os
+
+import pytest
+
+from fusion_engine_client.analysis.file_index import FileIndex
+from fusion_engine_client.messages import *
+from fusion_engine_client.parsers import FusionEngineEncoder, MixedLogReader
+from fusion_engine_client.utils.time_range import TimeRange
+
+
+class TestClass:
+    @pytest.fixture
+    def data_path(self, tmpdir):
+        data_path = tmpdir.join('test_file.p1log')
+        yield data_path
+
+    def _generate_data(self, data_path, message_types):
+        messages = []
+        prev_p1_time_sec = None
+        prev_system_time_sec = None
+        for message_cls, time_sec in message_types:
+            if isinstance(message_cls, bytes):
+                messages.append(message_cls)
+                continue
+
+            message = message_cls()
+            if hasattr(message, 'p1_time'):
+                if time_sec is None:
+                    if prev_p1_time_sec is None:
+                        message.p1_time = Timestamp()
+                    else:
+                        message.p1_time = Timestamp(prev_p1_time_sec)
+                else:
+                    message.p1_time = Timestamp(time_sec)
+                    if prev_p1_time_sec is not None:
+                        dt_sec = time_sec - prev_p1_time_sec
+                        prev_system_time_sec += dt_sec
+                    prev_p1_time_sec = time_sec
+            elif hasattr(message, 'system_time_ns'):
+                if time_sec is None:
+                    if prev_system_time_sec is None:
+                        message.system_time_ns = 0
+                    else:
+                        message.system_time_ns = int(prev_system_time_sec * 1e9)
+                else:
+                    message.system_time_ns = int(time_sec * 1e9)
+                    if prev_system_time_sec is not None:
+                        dt_sec = time_sec - prev_system_time_sec
+                        prev_p1_time_sec += dt_sec
+                    prev_system_time_sec = time_sec
+            messages.append(message)
+
+        encoder = FusionEngineEncoder()
+        with open(data_path, 'wb') as f:
+            for message in messages:
+                if isinstance(message, bytes):
+                    f.write(message)
+                else:
+                    f.write(encoder.encode_message(message))
+
+        return [m for m in messages if not isinstance(m, bytes)]
+
+    def _generate_mixed_data(self, data_path):
+        messages = self._generate_data(data_path, message_types=[
+            (EventNotificationMessage, 0.0),
+            (PoseMessage, 1.0),
+            (PoseMessage, 2.0),
+            (EventNotificationMessage, 2.0),
+            (PoseMessage, 3.0),
+            (PoseMessage, 4.0),
+            (EventNotificationMessage, 4.0),
+        ])
+        return messages
+
+    def _generate_mixed_data_with_binary(self, data_path):
+        messages = self._generate_data(data_path, message_types=[
+            (b'1234', None),
```

Comment:
Since this class uses some custom syncing logic, it might be worth having some of the "binary data" include the FE sync to exercise the logic of starting a sync, then needing to go back to resync after the failure is detecting.

Reply:
Yeah good point. I'll add that.